### PR TITLE
Merge 5.1.1minus 21

### PR DESCRIPTION
**Pull request by liam923**
The first two commits merge in flambda changes with conflicts and can be skipped.

@riaqn Could you review just the following commits?:
- 8e08c03a11fd35f094acf9c1265b26bc816fd8c4: Resolve closure_context build errors
- 148d6c53a03b513c3b3e0da6ae15c17d783e8010: Fix type errors related to modes and modalities
- d466b1e6b81209c5b733d6ed91c70d2297c9ba16: Fix more mode related things

### CODE REVIEW

File: src/analysis/ptyp_of_type.ml [hunk 1]
```diff
@@ -38,18 +38,24 @@ and core_type type_expr =
...
-      (core_type type_expr_out)
-      (failwith "TODO")
-      (failwith "TODO")
+  | Tarrow ((label,arg_alloc_mode,ret_alloc_mode), type_expr, type_expr_out, _commutable) ->
+    let (label : Asttypes.arg_label), type_expr = match label with
+      | Position l -> Labelled l, Typ.extension (mkloc "call_pos" !default_loc, PStr [])
+      | Nolabel -> Nolabel, core_type type_expr
+      | Labelled l -> Labelled l, core_type type_expr
+      | Optional l -> Optional l, core_type type_expr
+    in
```
Comment (liam923) [inline src/analysis/ptyp_of_type.ml:47]: These lines only changes in indentation

File: tests/test-dirs/type-enclosing/jane-street.t/run.t [hunk 2]
```diff
@@ -54,7 +54,7 @@ escape characters in string literals, so we use the revert-newlines script.
   type t3 : value = D | E of int
       ^
   With verbosity 0: "type t3 = D | E of int"
-  With verbosity 1: "type t3 = D | E of int"
+  With verbosity 1: "type t3 : value = D | E of int"
```
Comment (ncik-roberts) [inline tests/test-dirs/type-enclosing/jane-street.t/run.t:57]: In general, I find the diff in this file to be disappointing. The printing of layouts was specifically designed to not print `: value` unless the user wrote it. Do you know why it's now being printed spuriously?

Comment (liam923) [reply tests/test-dirs/type-enclosing/jane-street.t/run.t]: I've reverted this behavior

**Comment by liam923** [comment]
@ncik-roberts I just bumped the Menhir version to match the flambda version bump. As a result, the autogenerated file `src/ocaml/preprocess/parser_raw.ml` has a very large diff, which can be ignored.

File: tests/test-dirs/type-enclosing/jane-street.t/run.t (hunk 2 above)
Comment (ncik-roberts) [reply tests/test-dirs/type-enclosing/jane-street.t/run.t]: Thanks.

File: tests/test-dirs/type-enclosing/jane-street.t/run.t [hunk 3]
```diff
@@ -63,8 +63,8 @@ escape characters in string literals, so we use the revert-newlines script.
   
   type t5 : value
       ^
-  With verbosity 0: "type t5 : value"
-  With verbosity 1: "type t5 : value"
+  With verbosity 0: "type t5"
+  With verbosity 1: "type t5"
```
Comment (ncik-roberts) [inline tests/test-dirs/type-enclosing/jane-street.t/run.t:67]: There continues to be a diff in this test. `type t5 : value` should be printed here, as that's what the user wrote. Previously, the code checked the `jkind_annotation` in the typedtree to determine that.

**Review by ncik-roberts — APPROVED** [review]
This looks good to me -- just left one minor comment about the same kind-printing test.

**Pull request merged**